File: Data Structures & Algorithms/split-array-largest-sum/submission-2.java
class Solution {
    public int splitArray(int[] nums, int k) {
        int max = 0;
        int total = 0;
        for (int n: nums) {
            max = Math.max(max, n);
            total += n;
        }

        int l = max;
        int r = total;
        int res = 0;

        while (l <= r) {
            int m = l + ((r - l)/2);
            if (canSplit(nums, m, k)) {
                res = m;
                r = m - 1;
            } else {
                l = m + 1;
            }
        }

        return res;
    }

    private boolean canSplit(int[] nums, int expectedSum, int k) {
        int currSum = 0;
        int splits = 1;
        for (int n: nums) {
            if (currSum + n > expectedSum) {
                ++splits;
                currSum = n;
            } else {
                currSum += n;
            }
        }
        return splits <= k;
    }
}

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-2.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        pref_freq = {0 : 1}
        curr_sum = 0
        res = 0
        for n in nums:
            curr_sum += n
            diff = curr_sum - k
            if diff in pref_freq:
                res += pref_freq[diff]
            pref_freq[curr_sum] = pref_freq.get(curr_sum, 0) + 1
        return res
